Forge.TreeWalker/src/TreeNodeContext.cs

```diff
@@ -0,0 +1,100 @@
+//-----------------------------------------------------------------------
+// <copyright file="TreeNodeContext.cs" company="Microsoft">
+//     Copyright (c) Microsoft Corporation.  All rights reserved.
+// </copyright>
+// <summary>
+//     The TreeNodeContext class.
+// </summary>
+//-----------------------------------------------------------------------
+
+namespace Microsoft.Forge.TreeWalker
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// The TreeNodeContext class object is passed to Callbacks.BeforeVisitNode and Callbacks.AfterVisitNode.
+    /// </summary>
+    public class TreeNodeContext
+    {
+        /// <summary>
+        /// The unique identifier for this tree walking session.
+        /// </summary>
+        public Guid SessionId { get; private set; }
+
+        /// <summary>
+        /// The tree node key where the Action resides.
+        /// </summary>
+        public string TreeNodeKey { get; private set; }
+
+        /// <summary>
+        /// The dynamic properties of this Node given by the ForgeTree schema.
+        /// </summary>
+        public dynamic Properties { get; private set; }
+
+        /// <summary>
+        /// The dynamic user-defined context object that is able to be referenced when evaluating schema expressions or performing actions.
+        /// </summary>
+        public object UserContext { get; private set; }
+
+        /// <summary>
+        /// The cancellation token.
+        /// </summary>
+        public CancellationToken Token { get; private set; }
+
+        /// <summary>
+        /// The name of the ForgeTree in the JsonSchema.
+        /// </summary>
+        public string TreeName { get; private set; }
+
+        /// <summary>
+        /// The unique identifier for the root/parent tree walking session.
+        /// </summary>
+        public Guid RootSessionId { get; private set; }
+
+        /// <summary>
+        /// Indicates whether the tree walker should skip all actions defined in this node, and proceed with ChildSelector.
+        /// </summary>
+        /// <remarks>Allow Public set for testing purpose.</remarks>
+        public bool ShouldSkipActionsInTreeNode { get; set; }
```

Comment:
We actually would rely on the application to update this property. It will be decided inside BeforeVisitNode if we should skip the Actions.

Reply:
yes, only application and test cases should update this property. i added your comments to inline.

---
In reply to: [504223272](https://github.com/microsoft/Forge/pull/32#discussion_r504223272) [](ancestors = 504223272)

Reply:
"
When not null or whitespace, the tree walker will skip all actions defined in the current tree node, and proceed to AfterVisitNode then ChildSelector.
Update this property inside BeforeVisitNode if you wish to use this feature for the current tree node.
The string context is available to check in the current TreeNode's ChildSelector via Session.GetCurrentNodeSkipActionContext().
"

---
In reply to: [504257524](https://github.com/microsoft/Forge/pull/32#discussion_r504257524) [](ancestors = 504257524,504223272)

Reply:
updated the comment

---
In reply to: [505093484](https://github.com/microsoft/Forge/pull/32#discussion_r505093484) [](ancestors = 505093484,504257524,504223272)

Reply:
Please replace the entire summary with what I have above.

---
In reply to: [505150928](https://github.com/microsoft/Forge/pull/32#discussion_r505150928) [](ancestors = 505150928,505093484,504257524,504223272)

Reply:
"When set, the tree walker..."
I like that better.

---
In reply to: [505811771](https://github.com/microsoft/Forge/pull/32#discussion_r505811771) [](ancestors = 505811771,505150928,505093484,504257524,504223272)

Reply:
cool. fixed

---
In reply to: [505841676](https://github.com/microsoft/Forge/pull/32#discussion_r505841676) [](ancestors = 505841676,505811771,505150928,505093484,504257524,504223272)

Reply:
Can you remove the remarks? It isn't public just for testing.

---
In reply to: [505857375](https://github.com/microsoft/Forge/pull/32#discussion_r505857375) [](ancestors = 505857375,505841676,505811771,505150928,505093484,504257524,504223272)

Reply:
yes, in the latest design. it is not for testing only. Removed

---
In reply to: [505910307](https://github.com/microsoft/Forge/pull/32#discussion_r505910307) [](ancestors = 505910307,505857375,505841676,505811771,505150928,505093484,504257524,504223272)

Reply:
/// <summary>
        /// When set, the tree walker will skip all actions defined in the current tree node, and proceed to AfterVisitNode then ChildSelector.
        /// Update this property inside BeforeVisitNode if you wish to use this feature for the current tree node.
        /// The string context is available to check in the current TreeNode's ChildSelector via Session.GetCurrentNodeSkipActionContext().
        /// </summary>

---
In reply to: [505924425](https://github.com/microsoft/Forge/pull/32#discussion_r505924425) [](ancestors = 505924425,505910307,505857375,505841676,505811771,505150928,505093484,504257524,504223272)

Reply:
fixed. (i saw some duplication and unclear in the iteration #7)

---
In reply to: [506620610](https://github.com/microsoft/Forge/pull/32#discussion_r506620610) [](ancestors = 506620610,505924425,505910307,505857375,505841676,505811771,505150928,505093484,504257524,504223272)